public/app/features/dashboard-scene/scene/Scopes/ScopesFiltersBaseSelectorScene.tsx

```diff
@@ -0,0 +1,233 @@
+import { AppEvents, Scope, ScopeSpec, ScopeTreeItemSpec } from '@grafana/data';
+import { getAppEvents, getBackendSrv } from '@grafana/runtime';
+import { SceneObjectBase, SceneObjectState } from '@grafana/scenes';
+import { ScopedResourceServer } from 'app/features/apiserver/server';
+
+import { ScopesUpdate } from './events';
+
+export interface Node {
+  item: ScopeTreeItemSpec;
+  hasChildren: boolean;
+  isSelectable: boolean;
+  children: Record<string, Node>;
+}
+
+export interface ExpandedNode {
+  nodeId: string;
+  query: string;
+}
+
+export interface ScopesFiltersBaseSelectorSceneState extends SceneObjectState {
+  nodes: Record<string, Node>;
+  expandedNodes: ExpandedNode[];
+  scopes: Scope[];
+  isOpened: boolean;
+}
+
+const baseExpandedNode: ExpandedNode = {
+  nodeId: '',
+  query: '',
+};
+
+export abstract class ScopesFiltersBaseSelectorScene extends SceneObjectBase<ScopesFiltersBaseSelectorSceneState> {
+  private readonly serverGroup = 'scope.grafana.app';
+  private readonly serverVersion = 'v0alpha1';
+  private readonly serverNamespace = 'default';
+
+  protected readonly server = new ScopedResourceServer<ScopeSpec, 'Scope'>({
+    group: this.serverGroup,
+    version: this.serverVersion,
+    resource: 'scopes',
+  });
+
+  protected constructor(state: Partial<ScopesFiltersBaseSelectorSceneState> = {}) {
+    super({
+      nodes: {},
+      expandedNodes: [baseExpandedNode],
+      scopes: [],
+      isOpened: false,
+      ...state,
+    });
+
+    this.open = this.open.bind(this);
+    this.close = this.close.bind(this);
+    this.queryNode = this.queryNode.bind(this);
+    this.toggleNodeExpand = this.toggleNodeExpand.bind(this);
+    this.toggleNodeSelect = this.toggleNodeSelect.bind(this);
```

Comment:
Any particular reason not tu use arrow functions instead? I see you use bind a lot in different constructors.

Reply:
I personally hate mixing arrow-style methods with regular methods, hence I default to binding.